File: java/com/plasstech/lang/ssl/Lexer.java
package com.plasstech.lang.ssl;

public class Lexer {
  private final String text;

  private int loc;
  private char cc; // current character

  public Lexer(String text) {
    this.text = text;
    this.loc = 0;
    this.cc = 0;
    this.advance();
  }

  private void advance() {
    if (loc < text.length()) {
      cc = text.charAt(loc);
    } else {
      // Indicates no more characters
      cc = 0;
    }
    loc++;
  }

  public Token nextToken() {
    // skip unwanted whitespace
    while (true) {
      while (cc == ' ' || cc == '\n' || cc == '\t' || cc == '\r') {
        advance();
      }
      if (cc != '#') { // # comment
        break;
      }
      while (cc != '\n' && cc != '\r' && cc != 0) {
        advance();
      }
    }

    if (cc == 0) {
      return new Token(TokenType.EOF, "");
    }

    if (Character.isDigit(cc)) {
      return makeNumber();
    } else if (Character.isAlphabetic(cc)) {
      return makeText();
    } else if (cc == '"') {
      return makeString();
    }
    return makeSymbol();
  }

  private Symbol toSymbol(String symbolString) {
    for (Symbol symbolEnum : Symbol.values()) {
      if (symbolEnum.value.equals(symbolString)) {
        return symbolEnum;
      }
    }
    return null;
  }

  private Token makeSymbol() {
    var symbolString = String.valueOf(cc);
    advance(); // eat the first

    var maybeTwoCharSymbolString = symbolString + cc;
    var maybeTwoCharSymbol = toSymbol(maybeTwoCharSymbolString);
    if (maybeTwoCharSymbol != null) {
      advance(); // eat the second
      return new SymbolToken(maybeTwoCharSymbol);
    }

    var symbol = toSymbol(symbolString);
    if (symbol != null) {
      // NO ADVANCE HERE - we already advanced
      return new SymbolToken(symbol);
    }

    fail("Unknown symbol " + symbolString);
    return null;
  }

  private Token makeString() {
    advance(); // eat the "
    String value = "";
    while (cc != '"' && cc != 0) {
      value = value + cc;
      advance();
    }
    if (cc == '"') {
      advance();
    } else {
      fail("Expected closing double quote, found EOF");
    }
    return new ConstToken<String>(value, VarType.STR);
  }

  private void fail(String message) {
    throw new IllegalStateException(message);
  }

  private Token makeText() {
    var first = cc;
    advance();
    if (!Character.isAlphabetic(cc)) {
      String var = String.valueOf(first);
      if (first >= 'a' && first <= 'h') {
        return new VarToken(var, VarType.FLOAT);
      } else if (first >= 'i' && first <= 'n') {
        return new VarToken(var, VarType.INT);
      } else {
        return new VarToken(var, VarType.STR);
      }
    }
    String keyword = String.valueOf(first);
    try {
      while (Character.isAlphabetic(cc)) {
        keyword += cc;
        advance();
      }
      // look up the keyword
      var kw = Keyword.valueOf(keyword.toUpperCase());
      return new KeywordToken(kw);
    } catch (IllegalArgumentException e) {
      fail("Unknown keyword " + keyword);
      return null;
    }
  }

  private Token makeNumber() {
    var first = cc;
    advance();
    String num = String.valueOf(first);
    while (Character.isDigit(cc)) {
      num += cc;
      advance();
    }
    if (cc == '.') {
      num += cc;
      advance();
      while (Character.isDigit(cc)) {
        num += cc;
        advance();
      }
      return new ConstToken<Float>(Float.parseFloat(num), VarType.FLOAT);
    }
    return new ConstToken<Integer>(Integer.parseInt(num), VarType.INT);
  }
}

File: ruby/lexer.rb
require_relative "token"

class Lexer
  def initialize(text)
    @text = text
    @loc = 0
    advance()
  end

  def next_token
    while true
      while (not @cc.nil?) and @cc <= ' '
        advance()
      end
      if @cc != '#'
        break
      end
      # NOTE must use double quotes for backslash, sheesh. IHR
      while (not @cc.nil?) and @cc != "\n"
        advance()
      end
    end
    if @cc.nil?
      return Token.new(type: :eof)
    end
    if @cc =~ /[[:alpha:]]/
      return make_text()
    end
    if @cc =~ /[[:digit:]]/
      return make_number()
    end
    if @cc == '"'
      return make_string()
    end
    return make_symbol()
  end

  private 

  def make_symbol
    first = @cc
    val = first
    advance()
    if ['=', '<', '>', '!'].include?(first)
      # equals can follow
      if not @cc.nil? and @cc == '='
        val += '='
        advance()
      end
    end
    index = SYMBOL_STRINGS.index(val)
    if index.nil?
      fail("Unknown symbol #{val}")
    end
    sym = SYMBOLS[index]
    return Token.new(type: TokenType::Symbol, value: val, symbol: sym)
  end

  def advance
    @cc = @text[@loc]
    @loc = @loc + 1
  end

  def make_string
    advance()   # eat the opening quote
    val = ""
    while not @cc.nil? and @cc != '"'
      val += @cc
      advance()
    end
    if @cc.nil?
      fail "Unclosed string literal #{val}"
    end
    advance()
    return Token.new(type: TokenType::Constant, value: val, var_type: :string)
  end

  def make_number
    val = @cc
    advance()
    while not @cc.nil? and @cc =~ /[[:digit:]]/
      val += @cc
      advance()
    end
    if @cc != '.'
      return Token.new(type: TokenType::Constant, value: val, var_type: :int)
    end
    val += @cc
    advance()
    while not @cc.nil? and @cc =~ /[[:digit:]]/
      val += @cc
      advance()
    end
    return Token.new(type: TokenType::Constant, value: val, var_type: :float)
  end

  def make_text
    var = @cc
    advance()
    if not (@cc =~ /[[:alpha:]]/)
      lowervar = var.downcase #ihr
      if lowervar >= 'a' and lowervar <= 'h'
        return Token.new(type: TokenType::Variable, value: var, var_type: :float)
      elsif lowervar <= 'n' 
        return Token.new(type: TokenType::Variable, value: var, var_type: :int)
      else
        return Token.new(type: TokenType::Variable, value: var, var_type: :string)
      end
    end
    while not @cc.nil? and @cc =~ /[[:alpha:]]/
      var += @cc
      advance()
    end
    index = KW_STRINGS.index(var.upcase)
    if index != nil
      return Token.new(type: TokenType::Keyword, value: var, keyword: KEYWORDS[index])
    else
      fail("Unknown keyword " + var)
    end
  end
end

